### PR TITLE
Update migrator to populate `type` field on inline `Protocol` instances

### PR DESCRIPTION
**Pull request by brynnz22**
# Guidelines

## _Soft_ Schema Freeze

The `nmdc-schema` and `berkeley-schema-fy24` schemas are under a soft freeze, which means changes **should not** be made that have any downstream implications. To ensure this, all PRs created during the freeze will be closely reviewed with **every** component of the NMDC system in mind.

## Reviewers

To ensure no changes are made unexpectedly, PR creators will use this PR template to tag and notify all task coordinators. Review should be specifically requested from _all_ [Berkeley Schema Roll Out task coordinators](https://docs.google.com/document/d/1XXN1YuaBuSkxPXeiLKm5YxYzXTamBPQrzzeLhlh7PWs/edit#heading=h.u52g8v319adh) that you expect to be affected by this PR.

We expect task coordinators to review PRs and provide feedback/approval within 1 week of when they are identified as reviewers. 

PRs will **NOT** be merged until all task coordinators (or their delegates) have approved it; either here on GitHub (via "`Review changes` > `Approve`" or an equivalent comment) or verbally.

Expedition, questions, and discussion can happen at any meeting.

Delays in review & merging should be addressed in meetings or with NMDC leadership.

| If you expect the changes to<br>impact this component... | ...[request a review](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/requesting-a-pull-request-review)<br>from this person |
| -- | -- |
| Metadata<br>Schema | |
| Runtime<br>Mongo database<br>Database migrations | @eecavanna,<br>who will pull in<br>@shreddd as needed |
| Postgres<br>Ingest | |
| Data Portal | @aclum  |
| Workflows: MG & MT |  |
| Workflows: MetaB & NOM ||
| Workflows: LipidO | @kheal  |
| Workflows: MetaP | |
| ETL code |  |
| Jupyter notebooks | @brynnz22  |

# PR Information
 
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [ ] Feature
- [x] Bug Fix
- [ ] Optimization
- [ ] Documentation
- [ ] Schema change: Structure and content
  - created, updated, or deleted a `class`, `slot`, or `enum`
  - changed whether a `slot` is `multivalued`
  - changed the way a `slot` is assigned to a `class`
  - changed the `permissible_values` of an `enum`
  - _etc._
- [ ] Schema change: Cleanup and preparation
  - updated the description of a `class`, `slot`, or `enum`
  - updated the `mappings` of a `class`, `slot`, or `enum` to an ontology
  - added an `enum` for _future_ use (it is not in the `range` of any `slot`)
  - _etc._
     
## Description

> _PRs should be [small and concise](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/getting-started/best-practices-for-pull-requests#write-small-prs)._
>
This PR adds migrator to the currently existing migrator that updates the type slot. I just needed to move some lines of code around, but it was an easy fix. Also, when testing this migrator I noticed that the Part 17 migrator that specifically updates the `protocol_link` `type` errors out. I can show you how to test your migrators @kheal so you can be sure they work. Something to do with string and not lists. I am guessing some of those collections don't exist in Mongo (like the storage_process_set?) but I am not sure about this. I tested this updated migrator and it works. It updates the `protocol_link` slot properly.

## Related Issues

> All PRs should relate to or fix an issue(s). Please identify the issue(s) below.

- Related Issue(s): #
- Fixes: #

## Did you add/update any tests?

- [ ] Yes
- [ ] No _(Add a justification below)_
- [ ] I need help with writing tests

## Could this schema change make it so any valid data becomes invalid?

No

> This is a question about what the schema allows. It is not a question about what happens to exists in the NMDC database right now.
> 
> Example: If, in this PR branch, you renamed a `slot` from `foo` to `foo_bar`, the answer to this question would be "yes," **even if** nothing in the NMDC database _currently_ uses the `foo` `slot`.
>
> More examples: `slot` or `class` name changes, changes to a `slot`'s `multivalued` state, changes to a `slot`'s `range` (e.g. string to integer), changes to `slot` assignments to `class`es, changes to an `enum`'s `permissible_values`

- [ ] Yes _(A migrator is required)_
- [ ] No
- [ ] I need help determining this

## If you answered "Yes", does this PR branch include that migrator?
It edits an existing migrator
- [ ] Yes
- [ ] No, this PR is incomplete and I need help writing the migrator

## Does this PR have any downstream implications?

> Examples: any change here that requires a change to workflows, workflow automation, the Mongo-to-Postgres ingest process, Jupyter notebooks, the Runtime, etc.

- [ ] Yes _(Explain below)_
- [x] No

@eecavanna @kheal  can you be sure that all the collections this impacts are included? I used the part 17 migrator @kheal  as the guide for which collections it impacts (they already existed in this migrator, I just had to move them up to the inlined slots path in this migrator). I did not include the `collecting_biosamples_from_site_set` and `storage_process_set` because these collections are either empty in Mongo or don't actually exist in Mongo (e.g. `storage_process_set` was created in Berkeley and `collecting_biosample_from_site_set` is empty. The only subclasses of `MaterialProcessing` this impacts is `LibraryPreparation` and `Pooling` because they are what existed prior to Berkeley. 

@aclum I noticed that the MetatranscriptomeAnalysis is now three different classes (even in the NMDC schema). It looks like perlmutter is down for maintenance today so I can't see where that data exists. Previously this migrator added the `type` slot just in the`metatranscriptome_analysis_set` so I changed it to work with the three new collections since that collection no longer exists in the nmdc schema (but I am not able to test if they work on these collections because perlmutter appears to be down for maintenance and I am using a database pull from yesterday that did not include those collections. I assume they would work since the migrator works on the others. The new mongo collections are `metatranscriptome_annotation_set`, `metatranscriptome_assembly_set`, and `metatranscriptome_expression_analysis_set`. Just let me know if you foresee any problems with this.

### CODE REVIEW

File: nmdc_schema/migrators/partials/migrator_from_10_2_0_to_11_0_0/migrator_from_10_2_0_to_11_0_0_part_10.py
Comment (kheal) [inline nmdc_schema/migrators/partials/migrator_from_10_2_0_to_11_0_0/migrator_from_10_2_0_to_11_0_0_part_10.py:60]: `CollectingBiosamplesFromSite` (in collecting_biosamples_from_site_set) also has the slot `protocol_link` so we should include it here, I think?

The other classes I see with `protocol_link` are `MixingProcess`, `ChromatographicSeparationProcess`, and `SubsamplingProcess`, but there are no sets for that in nmdc-schema so I don't think we need to include them.

Comment (kheal) [reply nmdc_schema/migrators/partials/migrator_from_10_2_0_to_11_0_0/migrator_from_10_2_0_to_11_0_0_part_10.py:60]: I see your comment now about collecting_biosamples_from_site_set being empty in mongo, I'll leave it up to you to include or not.

**Comment by kheal** [comment]
I checked the classes that are affected by the `protocol_link` are included and I should have made it clear that I did not test the migrator in #250 against mongo, apologies all!

File: nmdc_schema/migrators/partials/migrator_from_10_2_0_to_11_0_0/migrator_from_10_2_0_to_11_0_0_part_10.py
Comment (eecavanna) [reply nmdc_schema/migrators/partials/migrator_from_10_2_0_to_11_0_0/migrator_from_10_2_0_to_11_0_0_part_10.py:60]: If the initial\* schema _allows_ for the `collecting_biosamples_from_site_set` collection to contain documents, I'd prefer it be included in this list. That way, if the collection is non-empty when this migrator gets run in the future, that collection gets migrated also.

Both the [do_for_each_document](https://github.com/microbiomedata/berkeley-schema-fy24/blob/b166c89665eb5a0703d401dd547008200df00db6/nmdc_schema/migrators/adapters/mongo_adapter.py#L211) and [set_field_of_each_document](https://github.com/microbiomedata/berkeley-schema-fy24/blob/b166c89665eb5a0703d401dd547008200df00db6/nmdc_schema/migrators/adapters/mongo_adapter.py#L198) adapter methods can handle non-existent collections, empty collections, and non-empty collections.

\* By "initial" schema, I mean the schema the database is being migrated _from_. In this case, the migrator is the 10th in a series of 16 that are used to migrate a database from the "legacy schema" to the "Berkeley schema." The 15 intermediate states of the database do not necessarily conform to any schema. I think the closest schema to an "initial" schema that we have in this context is the "[legacy](https://github.com/microbiomedata/nmdc-schema/blob/5930933d2115664ddcd94776f0d07201bb409960/src/schema/nmdc.yaml#L2910)" schema.

Comment (eecavanna) [inline nmdc_schema/migrators/partials/migrator_from_10_2_0_to_11_0_0/migrator_from_10_2_0_to_11_0_0_part_10.py:60]: This suggestion goes along with another one.

```suggestion
                                        "MetatranscriptomeExpressionAnalysis",
                                        "CollectingBiosamplesFromSite"
                                        ]
```

File: nmdc_schema/migrators/partials/migrator_from_10_2_0_to_11_0_0/migrator_from_10_2_0_to_11_0_0_part_10.py
Comment (eecavanna) [inline nmdc_schema/migrators/partials/migrator_from_10_2_0_to_11_0_0/migrator_from_10_2_0_to_11_0_0_part_10.py:99]: This suggestion goes along with another one.

```suggestion
            metatranscriptome_expression_analysis_set=[lambda document: self.add_type_slot_with_class_uri(document, "nmdc:MetatranscriptomeExpressionAnalysis", slots_with_inlined_classes)],
            collecting_biosamples_from_site_set=[lambda document: self.add_type_slot_with_class_uri(document, "nmdc:CollectingBiosamplesFromSite", slots_with_inlined_classes)]
        )  
```

**Review by eecavanna — APPROVED** [review]
Looks good to me—thanks for implementing this!

I added a couple sets of doctests that demonstrate the newly-implemented functionality.

I replied to an existing thread, [here](https://github.com/microbiomedata/berkeley-schema-fy24/pull/251#discussion_r1756116754), expressing my preference about including `CollectingBiosamplesFromSite` (class) and `collecting_biosamples_from_site_set` (collection) in the migrator. I also added GitHub-level suggestions that—if "accepted" (committed)—would add those things to the branch.

File: nmdc_schema/migrators/partials/migrator_from_10_2_0_to_11_0_0/migrator_from_10_2_0_to_11_0_0_part_10.py
Comment (brynnz22) [reply nmdc_schema/migrators/partials/migrator_from_10_2_0_to_11_0_0/migrator_from_10_2_0_to_11_0_0_part_10.py:88]: Okay I commited your suggestions @eecavanna and @kheal . CollectingBiosampleFromSite is included.

**Review by kheal — APPROVED** [review]

**Comment by eecavanna** [comment]
I'm tagging the Berkeley Schema Roll Out task coordinators here, so they are aware this PR exists. I'd prefer not to wait the full 1 week for the objection window to expire, given this PR's lack of downstream impact. I'll interpret a "👍" reaction to this message (as an alternative to the usual ways of communicating it) to mean you're OK with this PR being merged into `main`.

| If you expect the changes to<br>impact this component... | ...[request a review](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/requesting-a-pull-request-review)<br>from this person | Has<br>expressed<br>approval |
| -- | -- | -- |
| Metadata<br>Schema | @mslarae13 | |
| Runtime<br>Mongo database<br>Database migrations | @eecavanna,<br>who will pull  in<br>@shreddd as needed | Yes |
| Postgres<br>Ingest | @naglepuff | |
| Data Portal | @aclum | | 
| Workflows: MG & MT | @mbthornton-lbl | |
| Workflows: MetaB & NOM | @corilo | |
| Workflows: LipidO | @kheal | Yes |
| Workflows: MetaP | @SamuelPurvine | |
| ETL code | @sujaypatil96 | |
| Jupyter notebooks | @brynnz22 | Yes |

**Comment by eecavanna** [comment]
I received the "green light" on [Slack](https://nmdc-group.slack.com/archives/CFVH4DYGH/p1726524971800049?thread_ts=1726173528.448929&cid=CFVH4DYGH) to merge this in.

**Pull request merged**
